enhancements/shared-and-guaranteed-cpu-pools.md

```diff
@@ -0,0 +1,407 @@
+guaranteed CPUs.
+* The selection of shared vs. guaranteed CPUs should be transparent to the user - no changes to the pod
+spec should be required.
+* The configuration should be optional and activated with a feature gate.
+* Although the initial usecase for this enhancement is single-node deployments, it should also be usable
+for multi-node deployments.
+
+### Non-Goals
+
+* This enhancement assumes the feature is activated as part of installing the cluster and cannot be
+activated later. However, the sizing of the pools of shared and guaranteed CPUs can be modified (a reboot
+will be required).
+* The use of the shared and guaranteed CPU pools will be activated at the cluster level - it cannot be
+configured for a subset of the nodes in a cluster.
+
+## Proposal
+
+The [cluster-node-tuning-operator (NTO)](https://github.com/openshift/cluster-node-tuning-operator) will be
+extended to add a shared CPUSet to the [Performance Profile CPU configuration](https://github.com/openshift/cluster-node-tuning-operator/blob/master/docs/performanceprofile/performance_profile.md#cpu)
+in addition to the existing reserved and isolated CPUSets. For example:
+
+```yaml
+cpu:
+  reserved: 0,1
+  shared: 2-5
+  isolated: 6-15
+```
+When the shared CPUSet is configured, NTO will:
+* Continue to set the systemd.cpu_affinity to the reserved CPUSet.
+* Apply kernel configuration for isolated CPUs (e.g. nohz_full) to only the isolated CPUSet.
+* Update kubelet configuration to specify that the isolated CPUSet is to be used for guaranteed QoS
+containers using whole CPUs.
+* Update OpenShift Kubernetes API Server configuration to enable a new admission hook. Note: although this would
+work in single-node deployments, we still need to work through how enablement will work in multi-node deployments.
+
+The Kubelet configuration will be updated to allow a new CPUSet to be specified for guaranteed Qos containers
+using whole CPUs. There are two options:
+1. If this feature can be upstreamed, the [Kubelet Configuration](https://kubernetes.io/docs/reference/config-api/kubelet-config.v1beta1/) could be extended with a new `guaranteedCPUs` option.
+2. If this cannot be upstreamed, a new configuration file could be created and read by kubelet on startup - similar
+to the /etc/kubernetes/openshift-workload-pinning file created for The [Workload Partitioning](https://docs.openshift.com/container-platform/4.13/scalability_and_performance/ztp_far_edge/ztp-reference-cluster-configuration-for-vdu.html#ztp-sno-du-enabling-workload-partitioning_sno-configure-for-vdu) feature.
+
+The Kubelet (mostly the [CPU manager](https://github.com/openshift/kubernetes/tree/master/pkg/kubelet/cm/cpumanager) and its static policy) would be updated to:
+* Read and store the new `guaranteedCPUs` configuration.
+* Remove the `guaranteedCPUs` from the `defaultCpuSet` so that shared containers are excluded from the `guaranteedCPUs` when their affinity is set.
```

Comment:
@swatisehgal @ffromani I wonder if this change requires changes in the way of how RTE/NFD preform CPU accounting when running under OCP and this feature is enabled.

Reply:
@Tal-or , can you elaborate on how this would affect RTE/NFD ?

Reply:
RTE/NFD is not aware to the internal pool allocation and it might affect the accounting that TAS scheduler is doing.

Let's have an example (this is long bear with me):
Assuming I've a cluster with:
1. TAS running (RTE/NFD + scheduler) 
2.  shared-partition feature enabled
3.  topology-manager-policy=single-numa-node (in Kubelet under the nodes)

I also have `NodeA` with the following available/free resources layout (this is how Kubelet sees it):
```yaml
NUMA0:
  guaranteed-cpus: 1
NUMA1:
  guaranteed-cpus: 1
  shared-cpus: 1
```
  
A new guaranteed pod is coming requesting for 2 integral CPUs (guaranteed CPUs) and some memory.
After it goes through the mutation hook the pod spec looks as follow:
```yaml
spec:
  resources:
    requests:
      cpu: 2000m
      memory: 100Mi
      openshift.io/guaranteed-cpus: 2000
    limits:
      cpu: 2000m
      memory: 100Mi
      openshift.io/guaranteed-cpus: 2000
```

When the pod gets to the default scheduler the scheduler marks  `NodeA` as suitable for hosting the pod because it sees two available guaranteed-cpus (the default scheduler is not aware of NUMA locality).

All the nodes that marked as suitable by the default scheduler are examine later by TAS scheduler.
TAS scheduler marks `NodeA` as suitable for hosting the pod as well because it sees two available CPUs under NUMA1 (TAS scheduler fed by RTE/NFD information and RTE/NFD is not aware of garanteed/shared CPUs division)

Finally the pod admitted to `NodeA` and reject by Kubelet with `topologyAffinityError` (becuase there're no 2 guaranteed CPUs that affine to the same NUMA) which is exactly the kind of cases that TAS meant to solve.

Reply:
Thanks @Tal-or - that does sound like an issue. So what are the options for addressing it? Would this require changes to the RTE/NFD to add the new shared/guaranteed extended resources to the NodeResourceTopology? I assume there is another component that reports the actual usage of the resources per NUMA node? And then would the TAS scheduler itself have to be updated to take these new resources into account? Is there a reasonable way to make this kind of scheduling enhancement?

Reply:
So it seems we would need to publish the avail cpus per numa

Reply:
Changes in RTE are technically possible, but NFD is more complex.
NFD is the u/s version that intended to replace RTE. 
Since NFD is an u/s component, I'm not sure it's possible to incorporate the changes needed to support this design.

Reply:
I added a new "Scheduler Awareness" open question and a new "Enhance Topology Aware Scheduler" alternative to capture the issues you raised in your comments. I still need to investigate this new option.